witness/golang/omniwitness/usbarmory/internal/storage/persistance.go

```diff
@@ -0,0 +1,293 @@
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/witness/golang/internal/persistence"
+	"github.com/google/trillian-examples/witness/golang/omniwitness/usbarmory/internal/storage/slots"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	mappingConfigSlot = 0
+)
+
+// SlotPersistence is an implementation of the witness Persistence
+// interface based on Slots.
+type SlotPersistence struct {
+	// mu protects access to everything below.
+	mu sync.Mutex
+
+	// part is the underlying storage partition we're using to persist
+	// data.
+	part *slots.Partition
+
+	// mapSlot is a reference to the zeroth slot in a partition.
+	// This slot is used to maintain a mapping of log ID to slot index
+	// where state for that log is stored.
+	mapSlot *slots.Slot
+	// mapWriteToken is the token received when we read the mapping config
+	// from the mapSlot above. It'll be used when we want to store an updated
+	// mapping config.
+	mapWriteToken uint32
+
+	// idToSlot maintains the mapping from LogID to slot index used to store
+	// checkpoints from that log.
+	idToSlot slotMap
+
+	// freeSlots is a list of unused slot indices available to be mapped to logIDs.
+	freeSlots []uint
+}
+
+// slotMap defines the structure of the mapping config stored in slot zero.
+type slotMap map[string]uint
+
+// NewSlotPeristence creates a new SlotPersistence instance.
+// As per the Persistence interface, Init must be called before it's used to
+// read or write any data.
+func NewSlotPersistence(part *slots.Partition) *SlotPersistence {
+	return &SlotPersistence{
+		part:     part,
+		idToSlot: make(map[string]uint),
+	}
+}
+
+// populateMap reads the logID -> slot mapping from storage.
+// Must be called with p.mu locked.
+func (p *SlotPersistence) populateMap() error {
+	b, t, err := p.mapSlot.Read()
+	if err != nil {
+		return fmt.Errorf("failed to read persistence mapping: %v", err)
+	}
+	if err := yaml.Unmarshal(b, &p.idToSlot); err != nil {
+		return fmt.Errorf("failed to unmarshal persistence mapping: %v", err)
+	}
+	// We read the mapping config, so save the token for if/when we want to
+	// store an updated mapping.
+	p.mapWriteToken = t
+
+	// Precalculate the list of available slots.
+	slotState := make([]bool, p.part.NumSlots())
+	for _, idx := range p.idToSlot {
+		if idx == mappingConfigSlot {
+			return errors.New("internal-error, reserved slot 0 has been used")
+		}
+		slotState[idx] = true
+	}
+
+	// Slot 0 is reserved for the mapping config, so mark it used here:
+	slotState[mappingConfigSlot] = true
+
+	p.freeSlots = make([]uint, 0, p.part.NumSlots())
+	for idx, used := range slotState {
+		if !used {
+			p.freeSlots = append(p.freeSlots, uint(idx))
+		}
+	}
+	return nil
+}
+
+// storeMap writes the current logID -> slot map to storage.
+// Must be called with p.mu locked.
+func (p *SlotPersistence) storeMap() error {
+	smRaw, err := yaml.Marshal(p.idToSlot)
```

Comment:
IMO YAML is used where the config is going to be user edited, or at least, viewed. That said, it's probably fine here. It's just a bit of an odd choice for embedded storage, or at least, if you'd asked me in the pub how embedded stuff would persist things I probably wouldn't have mentioned YAML in my list of top 3 likely implementations.

Similarly for log state storage too.

I'm not saying this is wrong, I just want to make sure we've thought this through and are happy with it now, because it will be hard to change later.

Reply:
The thinking was that there's a trade off between storage space used and debug-ability if the wheels fall off,  I didn't really fancy dumping sectors and trying to piece together a GOB encoded blob so went for the eas(ier)-debug route given the greenness of this code.

Totally open to changing the encoding at a later date if the space overhead becomes unacceptable.

YAML in particular because we already have a dep on it due to config files.

Reply:
Aha, and there's the point I was trying to make. Changing the encoding at a later date is a bit of a tricky one because as soon as _anyone_ deploys this then we're stuck with the format unless we build a data migration tool in, or tell these witnesses to start over. Any witness doing this _seriously_ would need to rotate their keys at this point, which is quite an ecosystem ripple.

Reply:
Yeah, that's what I meant by "build a data migration tool in". It's just annoying because stuff like that you can't really remove because you don't know when it's all done. The joys of client software.

I'm happy to leave this as yaml for now, but let's add a README and note in there that this is only suitable for testing at the moment? Perhaps not having a README should be clue enough, but you know how it is :-)

Reply:
Added a skeleton README